src/plugins/dashboard/public/dashboard_api/get_dashboard_api.ts

```diff
@@ -7,47 +7,275 @@
+import {
+  getReferencesForControls,
+  getReferencesForPanelId,
+} from '../../common/dashboard_container/persistable_state/dashboard_container_references';
 import { initializeTrackPanel } from './track_panel';
 import { initializeTrackOverlay } from './track_overlay';
-import { initializeUnsavedChanges } from './unsaved_changes';
-
-export interface InitialComponentState {
-  anyMigrationRun: boolean;
-  isEmbeddedExternally: boolean;
-  lastSavedInput: DashboardContainerInput;
-  lastSavedId: string | undefined;
-  managed: boolean;
-  fullScreenMode: boolean;
-}
+import { initializeUnsavedChangesManager } from './unsaved_changes_manager';
+import { DASHBOARD_APP_ID, DEFAULT_DASHBOARD_INPUT } from '../dashboard_constants';
+import { LoadDashboardReturn } from '../services/dashboard_content_management_service/types';
+import { initializePanelsManager } from './panels_manager';
+import {
+  DASHBOARD_API_TYPE,
+  DashboardApi,
+  DashboardCreationOptions,
+  DashboardInternalApi,
+  DashboardState,
+} from './types';
+import { initializeDataViewsManager } from './data_views_manager';
+import { initializeSettingsManager } from './settings_manager';
+import { initializeUnifiedSearchManager } from './unified_search_manager';
+import { initializeDataLoadingManager } from './data_loading_manager';
+import { PANELS_CONTROL_GROUP_KEY } from '../services/dashboard_backup_service';
+import { getDashboardContentManagementService } from '../services/dashboard_content_management_service';
+import { openSaveModal } from './open_save_modal';
+import { initializeSearchSessionManager } from './search_session_manager';
+import { initializeViewModeManager } from './view_mode_manager';
+import { UnsavedPanelState } from '../dashboard_container/types';
+import { initializeTrackContentfulRender } from './track_contentful_render';
 
-export function getDashboardApi(
-  initialComponentState: InitialComponentState,
-  untilEmbeddableLoaded: (id: string) => Promise<unknown>
-) {
+export function getDashboardApi({
+  creationOptions,
+  incomingEmbeddable,
+  initialState,
+  initialPanelsRuntimeState,
+  savedObjectResult,
+  savedObjectId,
+}: {
+  creationOptions?: DashboardCreationOptions;
+  incomingEmbeddable?: EmbeddablePackageState | undefined;
+  initialState: DashboardState;
+  initialPanelsRuntimeState?: UnsavedPanelState;
+  savedObjectResult?: LoadDashboardReturn;
+  savedObjectId?: string;
+}) {
   const animatePanelTransforms$ = new BehaviorSubject(false); // set panel transforms to false initially to avoid panels animating on initial render.
-  const fullScreenMode$ = new BehaviorSubject(initialComponentState.fullScreenMode);
-  const managed$ = new BehaviorSubject(initialComponentState.managed);
-  const savedObjectId$ = new BehaviorSubject<string | undefined>(initialComponentState.lastSavedId);
+  const controlGroupApi$ = new BehaviorSubject<ControlGroupApi | undefined>(undefined);
```

Comment:
yes. control group API created by ReactEmbeddableRenderer and the API is returned via a callback. So its value is not available until later.

Reply:
If you did want to avoid this, you could potentially use an async function to wait for the control group to be created / assigned. I'd prefer to keep the subjects for items which are actually meant to change over the lifetime of the component.

Reply:
It makes sense to keep `controlGroupApi$` an observable to support controls as panels. `controlGroupApi$` will change as control panels are added and removed. Only change is that `controlGroupApi$` will return an array instead of a single value.

Reply:
I'm not completely sure that's the right way to think about the controls as panels project, but you're right that we'll need to revisit this in the future anyway, so it's not worth the effort to re-structure it right now.